pkg/operator/update/resources.go

```diff
@@ -0,0 +1,153 @@
+limitations under the License.
+*/
+
+package update
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+)
+
+func KeyspaceDiskSize(dst *planetscalev2.VitessKeyspaceTemplate, src planetscalev2.VitessKeyspaceTemplate) {
+	// Check that the keyspace definitions line up.
+	if !validateKeyspacePartitionings(*dst,src) {
+		return
+	}
+
+	updateDiskSize(dst, src)
+}
+
+func updateDiskSize(dst *planetscalev2.VitessKeyspaceTemplate, src planetscalev2.VitessKeyspaceTemplate) {
+	for i := range dst.Partitionings {
+		dstPartitioning := dst.Partitionings[i]
+		if dstPartitioning.Equal != nil {
+			updatePartitioningDiskSize(dstPartitioning.Equal, *src.Partitionings[i].Equal)
+		}
+	}
+}
+
+func updatePartitioningDiskSize(dst *planetscalev2.VitessKeyspaceEqualPartitioning, src planetscalev2.VitessKeyspaceEqualPartitioning) {
+	srcLoop:
+		for i := range dst.ShardTemplate.TabletPools {
+			dstTablet := dst.ShardTemplate.TabletPools[i]
+			var requestedTablet *planetscalev2.VitessShardTabletPool
+
+			for j := range src.ShardTemplate.TabletPools {
+				srcTablet := src.ShardTemplate.TabletPools[j]
+				// Match each dst tablet pool with its unique Type, Cell tuple in src.
+				if srcTablet.Type == dstTablet.Type && srcTablet.Cell == dstTablet.Cell {
+					requestedTablet = &srcTablet
+				}
+
+				// If we can't find a match, continue and try to find matches for other tablet pools.
+				if requestedTablet == nil {
+					continue srcLoop
+				}
+
+				if requestedTablet.DataVolumeClaimTemplate == nil {
+					continue srcLoop
+				}
+			}
+
+			srcRequests := &requestedTablet.DataVolumeClaimTemplate.Resources.Requests
+			dstRequests := &dstTablet.DataVolumeClaimTemplate.Resources.Requests
+
+			DiskResource(dstRequests, srcRequests)
+		}
+}
+
+func validateKeyspacePartitionings(dst planetscalev2.VitessKeyspaceTemplate, src planetscalev2.VitessKeyspaceTemplate) bool {
+	// Check that the list of partitionings are the same length.
+	if len(dst.Partitionings) != len(src.Partitionings) {
+		return false
+	}
+
+	for i := range dst.Partitionings {
+		// Validate that partitionings have the same type and shard count.
+		if !validatePartitionings(&dst.Partitionings[i], &src.Partitionings[i]) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func validatePartitionings(dst *planetscalev2.VitessKeyspacePartitioning, src *planetscalev2.VitessKeyspacePartitioning) bool {
+	if !validatePartitioningTypes(dst, src) {
+		return false
+	}
+
+	valid := false
+	if dst.Equal != nil {
+		valid = validateEqualPartitioning(dst.Equal, src.Equal)
+	} else {
+		valid = validateCustomPartitioning(dst.Custom, src.Custom)
+	}
+
+	return valid
+}
+
+func validatePartitioningTypes(dst *planetscalev2.VitessKeyspacePartitioning, src *planetscalev2.VitessKeyspacePartitioning) bool {
+	if dst.Equal != nil && src.Equal == nil {
+		return false
+	} else if dst.Equal == nil && src.Equal != nil {
+		return false
+	}
```

Comment:
We also need to check Custom partitionings. It's not guaranteed that one is always nil and one is always non-nil.

Reply:
It _should_ be the case that exactly one of them is non-nil, but we don't yet have a validating webhook to prevent invalid CRDs from being persisted, so we need to be liberal in what we accept.

Reply:
Gotcha!